File: algorithms/16_kotlin/graphs/Dijkstra.kt
package graphs

import java.util.PriorityQueue

data class Edge(val to: Int, val weight: Double)
data class NodeDist(val node: Int, val dist: Double) : Comparable<NodeDist> {
    override fun compareTo(other: NodeDist) = dist.compareTo(other.dist)
}

fun dijkstra(n: Int, adj: List<List<Edge>>, source: Int): DoubleArray {
    val dist = DoubleArray(n) { Double.POSITIVE_INFINITY }
    dist[source] = 0.0
    val pq = PriorityQueue<NodeDist>()
    pq.add(NodeDist(source, 0.0))

    while (pq.isNotEmpty()) {
        val (u, d) = pq.poll()
        if (d > dist[u]) continue
        for (edge in adj[u]) {
            if (dist[u] + edge.weight < dist[edge.to]) {
                dist[edge.to] = dist[u] + edge.weight
                pq.add(NodeDist(edge.to, dist[edge.to]))
            }
        }
    }
    return dist
}

fun main() {
    val adj = listOf(
        listOf(Edge(1, 4.0), Edge(2, 1.0)),
        listOf(Edge(3, 1.0)),
        listOf(Edge(1, 2.0), Edge(3, 5.0)),
        emptyList()
    )
    val d = dijkstra(4, adj, 0)
    assert(d[3] == 4.0)
    println("Kotlin Dijkstra verified.")
}

File: algorithms/13_swift/graphs/Dijkstra.swift
// Dijkstra's Algorithm in Swift (CLRS 3rd Ed. Chapter 24.3)

struct Edge {
    let to: Int
    let weight: Double
}

func dijkstra(n: Int, adj: [[Edge]], source: Int) -> [Double] {
    var dist = [Double](repeating: .infinity, count: n)
    var visited = [Bool](repeating: false, count: n)
    dist[source] = 0.0

    for _ in 0..<n {
        var u = -1
        for i in 0..<n {
            if !visited[i] && (u == -1 || dist[i] < dist[u]) {
                u = i
            }
        }
        if dist[u] == .infinity { break }
        visited[u] = true

        for edge in adj[u] {
            if dist[u] + edge.weight < dist[edge.to] {
                dist[edge.to] = dist[u] + edge.weight
            }
        }
    }
    return dist
}

let adj = [
    [Edge(to: 1, weight: 4.0), Edge(to: 2, weight: 1.0)],
    [Edge(to: 3, weight: 1.0)],
    [Edge(to: 1, weight: 2.0), Edge(to: 3, weight: 5.0)],
    []
]
let d = dijkstra(n: 4, adj: adj, source: 0)
assert(d[3] == 4.0)
print("Swift Dijkstra verified.")
